### PR TITLE
fix: fix the cache for g function

### PR DESCRIPTION
**Pull request by seriouszyx**
Signed-off-by: Yixiang Zhao <seriouszyx@foxmail.com>

### CODE REVIEW

**Comment by hsluoyz** [comment]
@nodece plz review

**Comment by nodece** [comment]
@seriouszyx `g()` is working now, but you also need to new a `g()` when enforce is called each time.

**Comment by hsluoyz** [comment]
@seriouszyx

**Comment by seriouszyx** [comment]
@nodece I'm confused. The enforce() implementation in Java is different from other languages. Should I just remove synchronized as below?

```java
private boolean enforce(String matcher, Object... rvals) {
        if (!enabled) {
            return true;
        }

        Map<String, AviatorFunction> functions = new HashMap<>();
        for (Map.Entry<String, AviatorFunction> entry : fm.fm.entrySet()) {
            String key = entry.getKey();
            AviatorFunction function = entry.getValue();

            functions.put(key, function);
        }
        if (model.model.containsKey("g")) {
            for (Map.Entry<String, Assertion> entry : model.model.get("g").entrySet()) {
                String key = entry.getKey();
                Assertion ast = entry.getValue();

                RoleManager rm = ast.rm;
                functions.put(key, BuiltInFunctions.generateGFunction(key, rm));
            }
        }
        aviatorEval = AviatorEvaluator.newInstance();
        for (AviatorFunction f : functions.values()) {
            aviatorEval.addFunction(f);
        }
        fm.setAviatorEval(aviatorEval);
        modelModCount = model.getModCount();

        String expString;
        if (matcher == null || "".equals(matcher)) {
            expString = model.model.get("m").get("m").value;
        } else {
            expString = Util.removeComments(Util.escapeAssertion(matcher));
        }
        ...
```

**Comment by nodece** [comment]
@seriouszyx It looks to use a global aviatorEval instance, it can improve performance, you don't should remove the synchronized, fight out a way to inject `g()` to this instance.

File: src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java [hunk 1]
```diff
@@ -57,9 +57,6 @@
     boolean autoNotifyWatcher = true;
     boolean autoNotifyDispatcher = true;
 
-    // cached instance of AviatorEvaluatorInstance
-    AviatorEvaluatorInstance aviatorEval;
-
     // detect changes in Model so that we can invalidate AviatorEvaluatorInstance cache
     int modelModCount;
```
Comment (nodece) [inline src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java:61]: Remove this

File: src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java [hunk 2]
```diff
@@ -399,37 +394,32 @@ private boolean enforce(String matcher, Object... rvals) {
...
+        }
+        AviatorEvaluatorInstance aviatorEval = AviatorEvaluator.newInstance();
+        for (AviatorFunction f : functions.values()) {
+            if (aviatorEval.containsFunction(f.getName())) {
+                aviatorEval.removeFunction(f.getName());
             }
+            aviatorEval.addFunction(f);
         }
+        fm.setAviatorEval(aviatorEval);
+        modelModCount = model.getModCount();
```
Comment (nodece) [inline src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java:422]: It is unnecessary.

File: src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java [hunk 3]
```diff
@@ -399,37 +394,32 @@ private boolean enforce(String matcher, Object... rvals) {
...
+            }
+        }
+        AviatorEvaluatorInstance aviatorEval = AviatorEvaluator.newInstance();
+        for (AviatorFunction f : functions.values()) {
+            if (aviatorEval.containsFunction(f.getName())) {
+                aviatorEval.removeFunction(f.getName());
             }
+            aviatorEval.addFunction(f);
         }
+        fm.setAviatorEval(aviatorEval);
```
Comment (nodece) [inline src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java:421]: It loos unnecessary.

**Comment by hsluoyz** [comment]
@nodece plz review

**Review by nodece — APPROVED** [review]
LGTM

**Comment by hsluoyz** [comment]
@seriouszyx what does "memorized does not work" mean? Can we change to a more understandable phrase?

**Comment by seriouszyx** [comment]
@hsluoyz Done

**Pull request merged**